graphql_api/helpers/connection.py

```diff
@@ -68,6 +69,110 @@ def page_info(self, *args, **kwargs):
         }
 
 
+class ArrayPaginator:
+    """Cursor-based paginator for in-memory arrays."""
+
+    def __init__(
+        self,
+        data: List[Any],
+        first: Optional[int] = None,
+        last: Optional[int] = None,
+        after: Optional[str] = None,
+        before: Optional[str] = None,
+    ):
+        self.data = data
+        self.start_index = 0
+        self.end_index = len(data)
+
+        if first and last:
+            raise ValueError("Cannot provide both 'first' and 'last'")
```

Comment:
use `ValidationError` instead so it can be handled well in the response, I have a feeling this will be an internal server error at worst or something ugly in the response for the user.